### ISSUE TITLE
Oat++ async endpoint auth

### ISSUE THREAD

**Issue by Maximsiv1410** [oatpp/oatpp#406]
Hello, first of all thx for this awesome framework!
My question is: Am I able to use authorization with ENDPOINT_ASYNC? I didn't find any examples of it's application with async controller and also my trials(same as with regular ENDPOINT) weren't successfull. Am i missing something or it's just not implemented yet?

**Comment by lganzzzo** [comment]
Hello @Maximsiv1410 ,

The convenience mappings (like `PATH`, `HEADER`, `AUTHORIZATION`, etc.) are available for `ENDPOINT` only.

If you are using `ENDPOINT_ASYNC` - then you have to implement everything manually.
To implement authorization in the Async endpoint you can use the authorization handler directly:

```cpp

oatpp::web::server::handler::BasicAuthorizationHandler authHandler("my_realm");

ENDPOINT_ASYNC(...) {

  Action act() {
    auto authHeader = request->getHeader(oatpp::web::protocol::http::Header::AUTHORIZATION);
    auto authObject = authHandler. handleAuthorization(authHeader);
    
    // TODO - use authObject
    // auth object should contain auth information
  
  }

};
```

Please also note, that it's recommended to use Simple API wherever possible - unless there is a clear need for Async APIs(in very rare cases).

More authorization examples you can find in oatpp tests:
- https://github.com/oatpp/oatpp/blob/master/test/oatpp/web/app/BasicAuthorizationController.hpp
- https://github.com/oatpp/oatpp/blob/master/test/oatpp/web/app/BearerAuthorizationController.hpp

**Comment by Maximsiv1410** [comment]
Thank you so much for your reply, your example code helped me a lot!

**Comment by Maximsiv1410** [comment]
Actually, I tried your example, but now I have next issue :
 HttpController.hpp:54:27: error: 'this' argument to member function 'handleAuthorization' has type 'const oatpp::web::server::handler::BearerAuthorizationHandler', but function is not marked const
AuthorizationHandler.hpp:197:40: note: 'handleAuthorization' declared here.
If all functions, called from async endpoint must be 'const' qualified(correct me if it is not), how can I use authHandler here?

**Comment by lganzzzo** [comment]
Hey @Maximsiv1410 ,

Yes, you are right. It would be better to have controller not `const`.
So for now, you just have to cast const away - a dirty hack :)

```cpp
class MyController : public oatpp::web::server::api::ApiController {
protected:
  typedef MyController __ControllerType;
private:
  oatpp::web::server::handler::BasicAuthorizationHandler m_authHandler;
protected:
  MyController(const std::shared_ptr<ObjectMapper>& objectMapper)
    : oatpp::web::server::api::ApiController(objectMapper)
    , m_authHandler("my-realm")
  {}
public:

  static std::shared_ptr<MyController> createShared(OATPP_COMPONENT(std::shared_ptr<ObjectMapper>, objectMapper)){
    return std::shared_ptr<MyController>(new MyController(objectMapper));
  }

  ENDPOINT_ASYNC("GET", "/", Root) {

    ENDPOINT_ASYNC_INIT(Root)

    Action act() override {
      auto authHeader = request->getHeader(oatpp::web::protocol::http::Header::AUTHORIZATION);

      /* cast const away - hack */
      auto authObject = std::static_pointer_cast<oatpp::web::server::handler::DefaultBasicAuthorizationObject>(
        ((MyController*)(controller))->m_authHandler.handleAuthorization(authHeader)
      );

      if(authObject->password == "123" && authObject->userId == "admin") {
        return _return(controller->createResponse(Status::CODE_200, "OK"));
      }

      return _return(controller->createResponse(Status::CODE_401, "Unauthorized"));

    }

  };
  
};
```

**Comment by lganzzzo** [comment]
Fixed in `1.3.0`

**Comment by doufu3344** [comment]
@lganzzzo , I found that if ENDPOINT_ASYNC needs to process the body data, if call `_return` before `readBodyToStringAsync()`, it will cause the next request data exception.
ENDPOINT_ASYNC:
```c++
    ENDPOINT_ASYNC("POST", "/body/string", EchoStringBody) {

        ENDPOINT_ASYNC_INIT(EchoStringBody)

        Action act() override {
            auto authHeader = request->getHeader(oatpp::web::protocol::http::Header::AUTHORIZATION);

            /* cast const away - hack */
            auto authObject = std::static_pointer_cast<oatpp::web::server::handler::DefaultBasicAuthorizationObject>(
                ((MyController*)(controller))->m_authHandler.handleAuthorization(authHeader)
                );

            if (authObject->password == "123" && authObject->userId == "admin") {
                /* return Action to start child coroutine to read body */
                return request->readBodyToStringAsync().callbackTo(&EchoStringBody::returnResponse);
            }

            return _return(controller->createResponse(Status::CODE_401, "Unauthorized"));

        }

        Action returnResponse(const oatpp::String & body) {
            /* return Action to return created OutgoingResponse */
            return _return(controller->createResponse(Status::CODE_200, body));
        }

    };
```
Testing request:
```shell
curl --location --request POST '127.0.0.1:8000/body/string' \
--header 'Content-Type: text/plain' \
--data-raw 'body'
```
The first request feedback after authentication failure:
```
server=oatpp/1.3.0
code=404
description=Not Found
message=No mapping for HTTP-method: 'bodyPOST', URL: '/body/string'
```
You can see that, the body content of the last failed request is appended before this request.

A solution is to perform authorization processing in the body data processing action:
```c++
    ENDPOINT_ASYNC("POST", "/body/string", EchoStringBody) {

        ENDPOINT_ASYNC_INIT(EchoStringBody)

        Action act() override {
            return _return(controller->createResponse(Status::CODE_401, "Unauthorized"));
        }

        Action returnResponse(const oatpp::String & body) {
            auto authHeader = request->getHeader(oatpp::web::protocol::http::Header::AUTHORIZATION);

            /* cast const away - hack */
            auto authObject = std::static_pointer_cast<oatpp::web::server::handler::DefaultBasicAuthorizationObject>(
                ((MyController*)(controller))->m_authHandler.handleAuthorization(authHeader)
                );

            if (authObject->password == "123" && authObject->userId == "admin") {
                /* return Action to start child coroutine to read body */
                return request->readBodyToStringAsync().callbackTo(&EchoStringBody::returnResponse);
            }

            /* return Action to return created OutgoingResponse */
            return _return(controller->createResponse(Status::CODE_200, body));
        }

    };
```
